Fix "remember me" checkbox on login

noahmoss opened the pull request:
Previously, the "remember me" checkbox was sending a `remember` field to the backend in the login request, but that was being entirely ignored. This PR changes the behavior so that permanent cookies are used when `remember me` is checked, and session cookies (which expire on closing the browser) are used when the box is unchecked. Setting the `MB_SESSION_COOKIES` env var will override the checkbox and force the use of session cookies for all logins.

Also updated the docs, and fixed some indentation in session.clj which is why a lot of lines are changed there.

I went down a bit of a rabbit hole trying to figure out if it would be possible to modify the login UI when `MB_SESSION_COOKIES` is set, to hide or gray out the "Remember me" checkbox since it won't have any effect. But I couldn't quite figure out how to access the environment variable at that point in the code. If anyone has any pointers on how to do this I can update this PR to include it.

Resolves #15223

noahmoss committed:
respect `remember` field on backend if session tokens are not enforced via env variable

noahmoss committed:
modify docs to describe new behavior

camsaul commented:
If you want to have the value of `MB_SESSION_COOKIES` available to the frontend, you just need to make it a Setting, e.g.

```clj
(defsetting session-cookies
  "Docstring goes here"
  :type :boolean
  :visibility :public
  :default false)
```

and then the FE will have access to it via `MetabaseSettings.get('session-cookies')`

camsaul commented on src/metabase/server/middleware/session.clj:
```diff
@@ -51,6 +51,15 @@
+    ;; Disallow permanent cookies if MB_SESSION_COOKIES is set
+    false
+    ;; Otherwise check whether the user selected "remember me" during login
+    (:remember (:body request))))
```
You can do `(get-in request [:body :remember])` here, or even `(-> request :body :remember)` -- I think those are a little cleaner, especially if you have > 2 keys. Not worth changing here tho unless you feel like it.

camsaul commented on src/metabase/server/middleware/session.clj:
```diff
@@ -65,29 +74,28 @@
-                                (deferred-trs "served over an insecure connection. Some browsers will reject ")
-                                (deferred-trs "cookies under these conditions. ")
-                                (deferred-trs "https://www.chromestatus.com/feature/5633521622188032")))))]
+                         {:same-site config/mb-session-cookie-samesite
```
The old indentation for `merge` was actually correct. `merge` is a function call, so if its arguments are all on new lines, they should be indented 1 space. 

2 spaces is for macros or special forms with a varargs `& body`, which includes things like `let` or `do` or `when` or `def`.

Your editor should do this for you automatically tho. Since `merge` is a core function it shouldn't even need to have an active nREPL connection to know the correct indentation. I'm not sure how to configure VIM do this correctly, but maybe @walterl can point you in the right direction

walterl replied:
This is actually one of the things that I haven't managed to figure out yet. Please let me know if you find a solution.

noahmoss replied:
@walterl I think I just got this behavior working by installing this plugin: https://github.com/guns/vim-clojure-static and putting `let g:clojure_align_subforms = 1` in my vimrc. Not sure if there are any unintended side effects of this plugin yet...

walterl replied:
Ah yes. I had that in my config for a while too, but it only fixes half the use cases: it either formats _all_ forms with one space (enabled), or _all_ forms with 2 spaces (disabled). The [single space indent guildeline](https://guide.clojure.style/#one-space-indent) is for _function calls_ with no args on the same line as the function. Macros taking a `& body` (e.g. `do`) should still be indented 2 spaces.

noahmoss replied:
Ah I see. Yeah, seems like `do` doesn't indent properly now.

noahmoss replied:
Here's a workaround to make specific words indent by two space: https://github.com/guns/vim-clojure-static#gclojure_fuzzy_indent-gclojure_fuzzy_indent_patterns-gclojure_fuzzy_indent_blacklist

`autocmd FileType clojure set lispwords+=do` makes `do` work properly for me.

walterl replied:
Great! I don't think I've tried that one before. :raised_hands:

camsaul commented:
It might be nice to add a test to `metabase.api.session-test` as well to make sure this works end-to-end (i.e. the `Set-Cookie` header should be different based on whether or not the API request has `:remember` in it). `mt/client-full-response` will give you the full response including the headers. You can probably just tweak the existing `login-test` a bit and have it test both with and without `:remember`

camsaul reviewed (changes requested):
Overall this LGTM, but a couple of minor things:

-  Old indentation was correct
- We should add an end-to-end test for `/api/session`
- Disabling that checkbox on the FE would be a nice to have, but I'm ok with that being a follow-on PR

noahmoss committed:
fix indentation

noahmoss committed:
use get-in to access remember field of request

noahmoss committed:
add a session api test

noahmoss committed:
expose session-cookies as a public setting

noahmoss committed:
hide remember me checkbox if session-cookies is set

noahmoss commented:
Addressed comments, and changed frontend to hide the checkbox when the `session-cookies` setting is `true`:

<img width="412" alt="Screen Shot 2021-05-13 at 5 09 37 PM" src="https://user-images.githubusercontent.com/32746338/118202356-9ce92900-b40e-11eb-8f50-e36930338824.png">

camsaul commented on src/metabase/server/middleware/session.clj:
```diff
@@ -51,6 +52,15 @@
   [response]
   (reduce clear-cookie (wrap-body-if-needed response) [metabase-session-cookie metabase-embedded-session-cookie]))
 
+(defn- use-permanent-cookies?
```
This might be a little misleading, since the cookies aren't permanent. They last 2 weeks or for however long someone configures session length. Maybe `use-persistent-cookies?` would be a better name?

camsaul reviewed (approved):
Cool, LGTM. One note about whether the `use-permanent-cookies?` function might be more accurately named if you change its name to `use-persistent-cookies?`. I'm cool with merging either way tho

noahmoss commented:
@camsaul I was actually going off of the MDN for the "permanent" terminology: https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#define_the_lifetime_of_a_cookie
